### PR TITLE
refactor: create a rust based `PartitionSet` 

### PR DESCRIPTION
**Pull request by universalmind303**
most of this is ported from the python impl inside `daft/runners/partitioning.py`. 

### Note for reviewer. 

For context around why this is needed. The `DataFrame` class uses `PartitionSet` extensively for various common operations such as `show`, and `collect`. In order to add this functionality to our spark connect implementation, we need a similar construct in rust. 

Ideally, I'd like to port over the python implementation to use this new rust one, but there are still a few things that I'm not entirely sure how to implement (such as `RayPartitionSet`)

Not all of the methods inside `partitioning.rs` are used yet, But I intend to follow up this PR with an implementation for https://github.com/Eventual-Inc/Daft/issues/3498, and this is a prerequisite as `show` relies on `get_preview_micropartitions`.

### CODE REVIEW

File: src/daft-connect/Cargo.toml [hunk 1]
```diff
@@ -12,7 +12,6 @@ daft-scan = {workspace = true}
 daft-schema = {workspace = true}
 daft-table = {workspace = true}
 dashmap = "6.1.0"
-derive_more = {workspace = true}
```
Comment (andrewgazelka) [inline src/daft-connect/Cargo.toml:15]: Is there a reason we don’t want to use `derive_more`?

File: src/daft-local-execution/src/pipeline.rs [hunk 2]
```diff
@@ -75,7 +75,7 @@ pub fn viz_pipeline(root: &dyn PipelineNode) -> String {
 
 pub fn physical_plan_to_pipeline(
     physical_plan: &LocalPhysicalPlan,
-    psets: &HashMap<String, Vec<Arc<MicroPartition>>>,
+    psets: &dyn PartitionSet,
```
Comment (andrewgazelka) [inline src/daft-local-execution/src/pipeline.rs:78]: What are your thoughts on making this `&impl (ParitionSet + ?Sized)`

File: src/daft-micropartition/src/partitioning.rs [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+        let partitions = self.values()?;
+        let mut partitions = partitions.into_iter().flat_map(|mp| mp.micropartitions());
+        partitions.try_fold(0, |acc, mp| Ok(acc + mp.size_bytes()?.unwrap_or(0)))
+    }
+
+    fn has_partition(&self, partition_id: &PartId) -> bool {
+        self.partitions.contains_key(partition_id)
+    }
+
+    fn delete_partition(&mut self, partition_id: &PartId) -> DaftResult<()> {
```
Comment (andrewgazelka) [inline src/daft-micropartition/src/partitioning.rs:185]: maybe add simple tests?

File: src/daft-micropartition/src/partitioning.rs [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+        Box::pin(futures::stream::iter(
+            self.partition.clone().into_iter().map(Ok),
+        ))
+    }
+}
+
+/// an arc'd reference to a [`PartitionBatch`]
+pub type PartitionBatchRef = Arc<dyn PartitionBatch>;
+
+pub trait PartitionSet {
```
Comment (andrewgazelka) [inline src/daft-micropartition/src/partitioning.rs:81]: please add docs to actual trait (thanks for method ones though)

File: src/daft-micropartition/src/partitioning.rs [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+        let size_bytes = mp.size_bytes().unwrap_or(None).unwrap_or(0);
+        Self {
+            num_rows,
+            size_bytes,
+            boundaries: None,
+        }
+    }
+}
+
+/// A batch of micropartitions
```
Comment (andrewgazelka) [inline src/daft-micropartition/src/partitioning.rs:35]:
```suggestion
/// A collection of related MicroPartitions that can be processed as a single unit.
/// Provides methods to access the underlying partitions either as a vector or a stream.
```

File: src/daft-micropartition/src/partitioning.rs [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+
+#[derive(Debug, Clone)]
+pub struct PartitionMetadata {
+    pub num_rows: usize,
+    pub size_bytes: usize,
+    pub boundaries: Option<Boundaries>,
+}
+
+impl PartitionMetadata {
+    pub fn from_micro_partition(mp: &MicroPartition) -> Self {
```
Comment (andrewgazelka) [inline src/daft-micropartition/src/partitioning.rs:24]: nittttttttt (super nit): any reason this is not a `impl From`

File: src/daft-micropartition/src/partitioning.rs [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+
+use common_error::{DaftError, DaftResult};
+use daft_dsl::Expr;
+use futures::stream::BoxStream;
+
+use crate::MicroPartition;
+type PartId = String;
+
+#[derive(Debug, Clone)]
+pub struct Boundaries {
```
Comment (andrewgazelka) [inline src/daft-micropartition/src/partitioning.rs:11]: Is boundaries the best name here? kinda confused why it is called boundaries

File: src/daft-micropartition/src/partitioning.rs [hunk 8]
```diff
@@ -0,0 +1,208 @@
+use std::{collections::HashMap, sync::Arc};
+
+use common_error::{DaftError, DaftResult};
+use daft_dsl::Expr;
+use futures::stream::BoxStream;
+
+use crate::MicroPartition;
+type PartId = String;
```
Comment (andrewgazelka) [inline src/daft-micropartition/src/partitioning.rs:8]:
```suggestion
type PartitionId = String;
```

I like not have abbreviations

Comment (universalmind303) [reply src/daft-micropartition/src/partitioning.rs:8]: most of the naming here was copied over from the python implementation

File: src/daft-micropartition/src/partitioning.rs (hunk 7 above)
Comment (universalmind303) [reply src/daft-micropartition/src/partitioning.rs:11]: The naming here was copied over from the python implementation. I wanted to make as much a 1-1 as possible.

File: src/daft-micropartition/src/partitioning.rs (hunk 6 above)
Comment (universalmind303) [reply src/daft-micropartition/src/partitioning.rs:24]: The naming here was copied over from the python implementation. I wanted to make as much a 1-1 as possible.

Comment (andrewgazelka) [reply src/daft-micropartition/src/partitioning.rs:24]: ok sgtm

File: src/daft-micropartition/src/partitioning.rs (hunk 7 above)
Comment (andrewgazelka) [reply src/daft-micropartition/src/partitioning.rs:11]: ok maybe link to it if possible in doc comment or mention it?

File: src/daft-micropartition/src/partitioning.rs (hunk 8 above)
Comment (andrewgazelka) [reply src/daft-micropartition/src/partitioning.rs:8]: hmmm ok 🥲 I guess it is fine... ideally we could rename both but I suppose that doesn't make this PR ⚛️

File: src/daft-connect/Cargo.toml (hunk 1 above)
Comment (universalmind303) [reply src/daft-connect/Cargo.toml:15]: we were only using it for `#[derive(Constructor)]` but we needed a custom impl now.

Comment (universalmind303) [reply src/daft-connect/Cargo.toml:15]: we can add it back when we need it again.

File: src/daft-micropartition/src/partitioning.rs (hunk 3 above)
Comment (universalmind303) [reply src/daft-micropartition/src/partitioning.rs:185]: right now there's quite a bit of setup for creating a micropartition in rust. I agree that we _should_ have tests, but our python implementation also didnt have any tests for this. So i think it should be fine for now.

Comment (andrewgazelka) [reply src/daft-micropartition/src/partitioning.rs]: ok sgtm

**Review by andrewgazelka — APPROVED** [review]
LGTM.

**Comment by andrewgazelka** [comment]
preference (specifically @universalmind303) on reviewers auto merging if review is requested and we ✅?

**Comment by universalmind303** [comment]
> preference (specifically @universalmind303) on reviewers auto merging if review is requested and we ✅?

i generally have a preference for leaving it to the author to merge.

**Pull request merged**